Multiplayer Client/Server Split Documentation (#2165)
o3de/o3de.org: content/docs/user-guide/networking/multiplayer/split.md

---
title: Client and Server Launchers with the Multiplayer Gem
description: Build Open 3D Engine launchers with the Multiplayer Gem specifically targeting Clients, Servers or both.
linkTitle: Separating Client and Server
---

The Multiplayer Gem supports builds that contain only client logic, only server logic, or both client and server logic. This allows users to create executables of smaller size by excluding unnecessary logic and dependencies. It also allows hiding potentially sensitive logic unique to one executable from the other. 

> * GameLauncher is a client only launcher

Reviewer: Nit: feels like this list needs an introduction ie with this split we now three runtime executables:

* ServerLauncher is a server only launcher suitable for Dedicated Servers

> * UnifiedLauncher provides both and is suitable for Client Hosted Servers

Reviewer: It it possible to turn off/not generate this target? ie if you never need a unified launcher (which I am assuming is the majority use case outside of testing)

Author: The target will always be generated as part of the VS solution but there is no requirement to build it. A user can opt not to build the UnifiedLauncher if they want.

These file lists are maintained in multiplayer_files.cmake and multiplayer_split_files.cmake respectively.

multiplayer_files.cmake generally contains core datatypes, base and core classes. multiplayer_split_files.cmake contains AutoComponent based MultiplayerComponents and types dependent on them.

### CMake setup

The split by cmake files leads us to four Multiplayer targets.

1. Common - A target containing multiplayer_files.cmake
2. Client - A target containing multiplayer_files.cmake plus multiplayer_split_files.cmake conditionally compiled for Clients
3. Server - A target containing multiplayer_files.cmake plus multiplayer_split_files.cmake conditionally compiled for Servers
4. Unified - A target containing multiplayer_files.cmake plus multiplayer_split_files.cmake conditionally compiled for Clients and Servers

When including the Multiplayer Gem it is important to understand the needs of said usage. If the usage requires split logic, it is recommended to create Client, Server and Unified targets which each use Multiplayer.Client, Server and Unified respectively. If it does not, Multiplayer.Common is sufficient.

> As an example, MultiplayerSample uses and builds upon MultiplayerComponents in the Multiplayer Gem. It therefore defines its own respective Client, Server and Unified targets. 

Reviewer: nit: link to MultiplayerSample project

---
title: Client and Server Launchers with the Multiplayer Gem
description: Build Open 3D Engine launchers with the Multiplayer Gem specifically targeting Clients, Servers or both.
linkTitle: Separating Client and Server
---

> The Multiplayer Gem supports builds that contain only client logic, only server logic, or both client and server logic. This allows users to create executables of smaller size by excluding unnecessary logic and dependencies. It also allows hiding potentially sensitive logic unique to one executable from the other. 

Reviewer 2: Would it be appropriate to provide an example of "sensitive logic"? "For example developers are encouraged to ensure their publicly released game client doesn't ship with server logic in order deter hacking."

Author: Sure, I don't see why not.

The splitting functionality produces multiple build types.
* GameLauncher is a client only launcher
* ServerLauncher is a server only launcher suitable for Dedicated Servers

> * UnifiedLauncher provides both and is suitable for Client Hosted Servers

Reviewer 2: So basically, if I ever want to switch from supporting only DedicatedServer to now supporting Client-Server, I need to stop giving players the ClientLauncher and instead give them the Unified Launcher?

Author: Practically speaking, that's probably how it would work out. A developer would likely want _any_ client to be able to host so they likely wouldn't use the ClientLauncher. They could still make use of the ServerLauncher for Dedicated hosting if they wanted.

Reviewer 3: Depends on a game type, an MMO might never used Unified but a regular co-op game would.

In the Multiplayer Gem's cmake, observe that each target enables or disables these traits based on the target. In example, Server enables `AZ_TRAIT_SERVER` while disabling `AZ_TRAIT_CLIENT`. Usage of these targets will bring the macro definitions with them.

{{< note >}}
The following CMake example is abbreviated
{{< /note >}}

```cmake
    ly_add_target(
        NAME Multiplayer.Client.Static STATIC
        NAMESPACE Gem
        FILES_CMAKE
            multiplayer_split_files.cmake
        COMPILE_DEFINITIONS
            PUBLIC
                AZ_TRAIT_CLIENT=1
                AZ_TRAIT_SERVER=0
        BUILD_DEPENDENCIES
            PUBLIC
                AZ::AzCore
                AZ::AzFramework
                AZ::AzNetworking
                Gem::Multiplayer.Common.Static
    )

    ly_add_target(
        NAME Multiplayer.Server.Static STATIC
        NAMESPACE Gem
        FILES_CMAKE
            multiplayer_split_files.cmake
        COMPILE_DEFINITIONS
            PUBLIC

>                 AZ_TRAIT_CLIENT=0

Reviewer 3: What happens in .Unified? Should user be setting both AZ_TRAIT_CLIENT and AZ_TRAIT_SERVER to 1?

Reviewer 3: Or, in this case it would be .Common target?

Author: That's correct, do you think it's worth adding the cmake for it?

Reviewer 4: +1 to adding an example for the unified (looks like it's already been added below)

---
title: Client and Server Launchers with the Multiplayer Gem
description: Build Open 3D Engine launchers with the Multiplayer Gem specifically targeting Clients, Servers or both.
linkTitle: Separating Client and Server
---

> The Multiplayer Gem supports builds that contain only client logic, only server logic, or both client and server logic. This allows users to create executables of smaller size by excluding unnecessary logic and dependencies. It also allows hiding potentially sensitive logic unique to one executable from the other. For example, a publicly released game client that doesn't ship with server logic can make hacking the simulation more difficult.

Reviewer: ```suggestion
The Multiplayer Gem supports separating code that contain only client logic, only server logic, or both client and server logic. This allows users to create executables of smaller size by excluding unnecessary logic and dependencies. It also allows hiding potentially sensitive logic unique to one executable from the other. For example, ensuring that a free-to-play client executable never includes any server logic code, to reduce the chances of hacking or abuse.
```

> The splitting functionality produces multiple build types.

Reviewer: ```suggestion
The splitting functionality produces multiple build types:
```

* GameLauncher is a client only launcher
* ServerLauncher is a server only launcher suitable for Dedicated Servers

> * UnifiedLauncher provides both and is suitable for Client Hosted Servers

Reviewer: nit: A unified launcher is a bit of a weird beast, and IMHO we should cover the use case for this a little more.

This functionality is implemented through a variety of build mechanisms and it's important to understand these mechanisms in any Gem or project using the Multiplayer Gem.

## Splitting Up Client and Server logic

> The Multiplayer Gem contains code files that can be divided into two categories.

Reviewer: ```suggestion
The Multiplayer Gem contains code files that can be divided into two categories:
```

1. Files that are fully required on all launcher types.
2. Files that have parts conditionally compiled out depending on launcher type and their dependents.

> These file lists are maintained in multiplayer_files.cmake and multiplayer_split_files.cmake respectively.

Reviewer: ```suggestion
These file lists are maintained in `multiplayer_files.cmake` and `multiplayer_split_files.cmake` respectively.
```

Would a developer call these files differently? Or are these specific to the Multiplayer Gem?

Author: They are specific to the Multiplayer Gem and would only be a consideration for a developer if they are specifically modifying the Multiplayer Gem. Otherwise they would use one of Multiplayer's targets (Common, Client, Server or Unified).

> multiplayer_files.cmake generally contains core datatypes, base and core classes. multiplayer_split_files.cmake contains AutoComponent based MultiplayerComponents and types dependent on them.

Reviewer: ```suggestion
`multiplayer_files.cmake` generally contains core datatypes, base and core classes. `multiplayer_split_files.cmake` contains AutoComponent based MultiplayerComponents and types dependent on them.
```

### CMake setup

> The split by cmake files leads us to four Multiplayer targets.

Reviewer: ```suggestion
The split by cmake files leads us to four Multiplayer targets:
```

1. Common - A target containing multiplayer_files.cmake
2. Client - A target containing multiplayer_files.cmake plus multiplayer_split_files.cmake conditionally compiled for Clients
3. Server - A target containing multiplayer_files.cmake plus multiplayer_split_files.cmake conditionally compiled for Servers

> 4. Unified - A target containing multiplayer_files.cmake plus multiplayer_split_files.cmake conditionally compiled for Clients and Servers

Reviewer: Do we provide any guidance in how to ignore avoid generating a unified target? Imagine a lot use cases will be pure client/server split.

Author: The Unified target is needed in a few places (such as the Editor) so that's unfortunately not the case. The user is however under no obligation to build or use the UnifiedLauncher. The build target will exist but is in no way required.

> 4. Unified - A target containing multiplayer_files.cmake plus multiplayer_split_files.cmake conditionally compiled for Clients and Servers

Reviewer 4: ```suggestion
4. Unified - A target containing multiplayer_files.cmake plus multiplayer_split_files.cmake conditionally compiled for both Clients and Servers
```
